mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -137,14 +139,17 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
 
-      // Compute threshold to get exact k elements.
-      // TODO: SPARK-18409: Use approxQuantile to get the threshold
-      val modelDatasetSortedByHash = modelDataset.sort(hashDistCol).limit(numNearestNeighbors)
-      val thresholdDataset = modelDatasetSortedByHash.select(max(hashDistCol))
-      val hashThreshold = thresholdDataset.take(1).head.getDouble(0)
-
-      // Filter the dataset where the hash value is less than the threshold.
-      modelDataset.filter(hashDistCol <= hashThreshold)
+      // Compute threshold to get around k elements.
+      var approxQuantile = numNearestNeighbors.toDouble / count + 0.05  // relative error = 0.05
```

Comment:
Nit: can this be `val`? also do you want a `val relativeError = 0.05` and reuse it?
Might also be worth a short comment saying that the "err + M/N" quantile should be guaranteed to give enough neighbors.